design/0001-metallb-crd.md

```diff
@@ -0,0 +1,251 @@
+# MetalLB Custom Resource Controller
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Initial CRD definition](#initial-crd-definition)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to be able to use kubernetes Custom
+Resource Definition "CRD" to configure MetalLB layer2 and BGP.
+
+## Motivation
+
+Today MetalLB allows only configuration of layer2 and BGP protocols using
+ConfigMap objects, which is simple from configuration perspective and doesn't
+require cluster-admin rights, however it doesn't have validation on the API
+server level, no support for API versioning and no fine-grained permission
+model via RBAC.
+Custom resource definition "CRD" takes advantage of kubernetes extensibility
+to deliver the automation advantages of cloud services like provisioning,
+scaling and backup/restore.
+
+### Goals
+
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD"
+instead of ConfigMap.
+
+## Proposal
+
+Using SIG's [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+framework to build new CRD "MetalLB" object and a controller which works as
+a translator from the user's configuration yaml to a MetalLB's configmap
+object, with this approach there will be no changes required for the
+MetalLB management plane.
+
+### Risks and Mitigations
+
+Users need to either use CRD or ConfigMap to provision load balancer instance
+but not both, if user mistakenly used both the CRD configuration will
+overwrite the user defined ConfigMap.
+
+## Design Details
+
+![Alt text](./metallb_crd.png?raw=true "MetalLB Kubernetes Controller")
+
+1. User will supply the yaml configuration using the MetalLB resource.
+1. Configuration will get authenticated and authorized based on RBAC rules.
+1. Supplied configuration will get verified using openAPIV3Schema based on
+   the definition of MetalLB CRD.
+1. New objects will be stored in etcd.
+1. MetalLB controller monitors API server looking for creation of MetalLB
+   resource.
+1. MetalLB controller will use golang based template to convert the configured
+   MetalLB object to ConfigMap object.
+1. The generated ConfigMap object will be written to etcd so MetalLB's
+   components can consume it as if the user directly supplied the ConfigMap
+   configuration.
+
+### Initial CRD definition
+
+In this section will share initial metallB custom resource definition which
+currently covers only layer2, BGP will be added to the same CRD, or we might
+separate it out and use the same controller to manage both resources.
+
+Note: the current definition includes NodeAffinity configuration which not
+upstream yet, but we have it here for completeness.
```

Comment:
If you're going to leave that in, I would like to have it in a separate section that gives an illustration of the process of extending it to support new configuration.  I think we should specify what we'd expect to implement and merge by itself, but having a real example of how it would be extended would also be helpful.  That will be important to understand.

Reply:
that will be doc in the next PR with the rest of the operator code changes